### DISCUSSION TITLE
Cannot Distribute Mac Builds (Hardened Runtime)

### DISCUSSION THREAD

**Discussion by TyelorD** [axmolengine/axmol#1810] [Q&A] [answered]
Upon successfully making an Archive build with Xcode 15.2, then under Organizer I select that archive and press Distribute App.

Selecting Direct Distribution, and then Distribute gives me the following error:

```
Hardened Runtime is Not Enabled
"[AxmolGame].app" must be rebuilt with support for the Hardened Runtime. Enable the Hardened Runtime capability in the project editor, test your app, rebuild your archive, and upload again.
```

However, under Signing and Capabilities, I do have Hardened Runtime enabled, as well as under the Build Settings for each target I have "Enable Hardened Runtime" set to "Yes". I've also set the Code Signing Entitlements to our entitlements file, the Code Signing Identity to "Apple Development", the Code Signing Style to "Automatic", and the Development Team to the Development Team I work for.

This is setup exactly how I had it setup for Cocos2d-x 3.17.2, except that Cocos2d-x didn't give us any errors like this upon attempting to distribute an archive build.

It seems like either Xcode isn't signing or attempting to sign our build, so the Archive itself is unsigned: thus giving us the Hardened Runtime missing error. I'm not sure why this might be though, and don't see anything in the project file that would stop Xcode from signing the build.

Any help solving this issue would be greatly appreciated!

**Comment by rh101** [comment]
A similar issue was discussed in #1662 and resolved.  Please take a look at that discussion to see if it helps.

**Reply by TyelorD** [reply]
I should've mentioned that I've already looked into #1662. I don't believe I'm having the same issue there, and I'm not using CMake to build anymore after calling axmol build once to generate the project files. (I don't have the authority to change the company's entire build and release pipeline for our games, so I have to continue to build with Xcode, Android Studio, and Visual Studio rather than the CLI tools Axmol provides.)

Under Signing and Capabilities, I do have Hardened Runtime enabled, as well as under the Build Settings for each target I have "Enable Hardened Runtime" set to "Yes". I've also set the Code Signing Entitlements to our entitlements file, the Code Signing Identity to "Apple Development", the Code Signing Style to "Automatic", and the Development Team to the Development Team I work for.

Nothing there gave me any errors or warnings like the user who reported #1662 had, so I assume this means it's setup correctly in the project files?

The link you suggested would only set those settings in the project file for me, which I've already done. Unless I'm misunderstanding what the that CMake flag is doing upon building.

I've already spent a lot of time manually editing the xcproject file so that it builds outside of my local machine when the latest depot is gotten from P4v, rather than pointing to files locally on my machine, as well as some setup required to get our game running. I really don't want to remake all of these changes again.

P.S. This does raise the question of why the signing is explicitly disabled by default for macOS builds considering that makes it impossible to distribute those macOS builds anywhere. You cannot submit that build to the Apple App Store, or make a Direct Distribution for that build; meaning you cannot distribute that build anywhere outside of your local machine as far as I'm aware.

**Reply by TyelorD** [reply, marked as answer]
I apologize, I didn't realize that CODE_SIGNING_ALLOWED was a User-Defined project build setting in the xcproject file, and assumed it was a setting under "Signing".

After changing this from "NO" to "YES", I was able to build an archive and distribute it with Direct Distribution from the Xcode Organizer window.
